Pm 4 frontend setup

asching7108 opened the pull request:
Frontend setup include React routes, bootstrap, and a basic nav bar. A few notes:

- Frontend project structure under /src:
  - routes: pages
  - components: partial components (eg: App, Header/nav bar)
  - context: variables or methods we want to use across pages, eg: login status
  - services: handles api calls and provides functions for routes/components/context

- /components/App/App.js:
  - handles routes
  - includes public (if logged in, redirect to home page) and private (if not logged in, redirect to login page) routes

- setup bootstrap

- setup nav bar

- Please run `npm i` for newly added packages and `npm start` to test

asching7108 committed:
PM-4 setup React, Bootstrap and Navbar

asching7108 committed:
PM-4 removed unused code and changed function name

xHoudek commented on src/components/Header/Header.js:
```diff
@@ -0,0 +1,62 @@
+    this.context.setShelterAdminState(false);
+  }
+
+  renderLoginLink() {
```
Genuine question: Should this be combined with renderLogoutLink() and have the code logic here, instead of the render at line 36? I don't know enough about React to determine which one would be better in the long run. What are your thoughts?

```
determineLoginLink() {
    this.context.isLoggedIn
        ? return(<Link>LogOut</Link>)
        : return(<Link>LogIn</Link>)
}
```

asching7108 replied:
I think we can do it either way and I honestly can't tell which is better practice. I separated renderLoginLink() and renderLogoutLink() because we need a handler only for the logout, so I felt it's more clear to separate them. (If you click logout, you're logged out immediately so we need to call the API and update the context, but login only takes you to the login page and the rest will be handled there.)

xHoudek replied:
Alright in that case I think we can leave it as-is and refactor later if we want!

xHoudek commented on src/components/App/App.js:
```diff
@@ -0,0 +1,54 @@
+          <Routes>
+            <Route path={'/'} element={<HomePage />} />
+            <Route
+              path="/login"
```
Not sure if I am doing something wrong, but I was unable to access this page

asching7108 replied:
It's because in the AuthService I have this temporary method implementation:
```
isLoggedIn() {
    // TODO
    return true;
  },
```
Since you're logged in and the login page is using PublicRoute, the app will redirect you to the home page.
On the other hand, if you change the isLoggedIn() method to return false, you can access the login page, but if you click on the favorites page you'll be redirected to the login page.

xHoudek replied:
Makes sense to me. Sounds like that will work great. Thanks for the clarification!

xHoudek reviewed (changes requested):
Looks good from what I can tell! My limited knowledge of React means that I don't know exactly what is going on in each thing you added, but it looks like a good amount was put into it! A couple of thoughts:

1. We didn't really talk about naming convention at all. Is there a reason that components like Header are capitalized? Is PascalCase standard in React? Otherwise I would suggest using camelCase instead. Similar comment to auth-service.js. Unless there is a reason for the dash, I would suggest following a naming convention.
2. GitHub is mad about the lack of newlines at the end of files. I would suggest adding them to files just to make it happy
3. See comments about specific things

After we sort out the naming convention and newline stuff, I think this PR is good to go!

asching7108 committed:
Added newline at end of files and rename file

asching7108 commented:
Thank you for doing a thorough code review Alex!
I did some research on React naming convention, it seems there's no official guideline, but this Airbnb practice seems to be widely adopted:
https://github.com/airbnb/javascript/tree/master/react#naming
The [create-react-app](https://create-react-app.dev/docs/folder-structure/) also default to using PascalCase for components (I guess routes and context are considered components too).

Besides that, I agree that for index.js (root component) and other services files, we should use camelCase.

I added newlines at end of files and changed auth-service.js to authService.js.

xHoudek reviewed (approved):
Looks good!
